### ISSUE TITLE
Charged phases that cannot charge balance raise errors

### ISSUE THREAD

**Issue by bocklund** [pycalphad/pycalphad#503]
### Discussed in https://github.com/pycalphad/pycalphad/discussions/502

<div type='discussions-op-text'>

<sup>Originally posted by **zliang1990** December 13, 2023</sup>
Hi Richard (@richardotis),

My name is Zhi Liang, Materials Design Engineer in QuesTek. When I recently attempted to run the attached V-O description (GitHub does not support .TDB so I zipped it) with Pycalphad, I noticed that the following error message will pop up if I only choose oxygen (and vacancy) as elements.

Upon debugging, it seems to do with the HALITE phase, whose sublattice model is V,V+2,V+3,VA : O-2,VA. When only choosing oxygen, first sublattice only allows VA while second sublattice allows O-2 and VA, which seemingly triggers the following error message. This issue can be a blocker if in a multicomponent system, certain phase only has sublattice model A,VA : O-2,VA, but the element selection only has element B and oxygen, and this error message will pop up as well (Sorry that I cannot provide our multicomponent database since it is proprietary).

Could you please help to see how we can resolve this?

Thanks!

Best,
Zhi

![image](https://github.com/pycalphad/pycalphad/assets/84593259/fce335ce-9037-4f6c-8a72-e112d55c2a3e)
[15YANG_V-O description.zip](https://github.com/pycalphad/pycalphad/files/13666266/15YANG_V-O.description.zip)</div>

**Comment by bocklund** [comment]
From #502:

What's happening here is that when only vacancies and oxygen are entered as elements PyCalphad is treating CORUNDUM and HALITE as active phases (because there are active species in each sublattice).

For both phases, it's impossible to charge balance given these components (pure vacancy endmembers don't count, such as in the case for HALITE) and that gives array size errors because there are no feasible points. I'm creating an issue for this, since it's a bug. We should probably prevent phases that cannot charge balance from being included in the active set of phases.

**Comment by richardotis** [comment]
This should be easy to fix with the new sampler, as it only samples the feasible subspace of points. For these cases it would find no solution, so it would be easy to raise an error (or return zero points) so the phase is automatically excluded from the active set.

**Comment by richardotis** [comment]
This is now fixed in #432

**Comment by nury12n** [comment]
I noticed the test (pycalphad/tests/test_work_space.py::test_issue_503_pure_vacancy_charge_balance) for this issue can occasionally fail.

These were some of the outputs I got when the test failed:
In the first case, the phase fraction of the GAS phase was nan, and in the second case, there was a zero division error in the minimizer.
```
====================================================== FAILURES =======================================================
_____________________________________ test_issue_503_pure_vacancy_charge_balance ______________________________________

    def test_issue_503_pure_vacancy_charge_balance():
        "Pure vacancy phases are correctly suspended (gh-503)"
        TDB = """
        ELEMENT /-   ELECTRON_GAS              0.0000E+00  0.0000E+00  0.0000E+00!
        ELEMENT VA   VACUUM                    0.0000E+00  0.0000E+00  0.0000E+00!
        ELEMENT O    1/2_MOLE_O2(G)            1.5999E+01  4.3410E+03  1.0252E+02!
        ELEMENT ZR   BLANK                     0.0000E+00  0.0000E+00  0.0000E+00!
        SPECIES O-2                         O1/-2!
        SPECIES ZR+4                        ZR1/+4!
        PHASE SPINEL %  4 1 2 2 4 !
        CONSTITUENT SPINEL : ZR+4 : VA : VA : O-2 :  !
        PHASE GAS:G %  1  1.0  !
        CONSTITUENT GAS:G :O,ZR :  !
        """
        wks = Workspace(TDB, ['O', 'ZR', 'VA'], ['SPINEL', 'GAS'], {v.P: 1e5, v.X('O'): 1, v.T: 1000})
        assert np.isnan(wks.get('NP(SPINEL)'))
>       np.testing.assert_almost_equal(wks.get('NP(GAS)'), 1.0)

pycalphad\tests\test_workspace.py:284:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\ury3\AppData\Local\Programs\Python\Python39\lib\contextlib.py:79: in inner
    return func(*args, **kwds)
C:\Users\ury3\AppData\Local\Programs\Python\Python39\lib\contextlib.py:79: in inner
    return func(*args, **kwds)
..\..\..\..\pyvenv\pycalphad-workspace\lib\site-packages\numpy\_utils\__init__.py:85: in wrapper
    return fun(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<function assert_array_almost_equal.<locals>.compare at 0x000002168F899D30>, array(nan), 1.0)
kwds = {'err_msg': '', 'header': 'Arrays are not almost equal to 7 decimals', 'precision': 7, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError:
E           Arrays are not almost equal to 7 decimals
E
E           nan location mismatch:
E            ACTUAL: array(nan)
E            DESIRED: array(1.)

C:\Users\ury3\AppData\Local\Programs\Python\Python39\lib\contextlib.py:79: AssertionError
=============================================== short test summary info ===============================================
FAILED pycalphad/tests/test_workspace.py::test_issue_503_pure_vacancy_charge_balance - AssertionError:
=========================== 1 failed, 264 passed, 2 skipped, 1 xfailed in 74.51s (0:01:14) ============================
```

```
====================================================== FAILURES =======================================================
_____________________________________ test_issue_503_pure_vacancy_charge_balance ______________________________________

    def test_issue_503_pure_vacancy_charge_balance():
        "Pure vacancy phases are correctly suspended (gh-503)"
        TDB = """
        ELEMENT /-   ELECTRON_GAS              0.0000E+00  0.0000E+00  0.0000E+00!
        ELEMENT VA   VACUUM                    0.0000E+00  0.0000E+00  0.0000E+00!
        ELEMENT O    1/2_MOLE_O2(G)            1.5999E+01  4.3410E+03  1.0252E+02!
        ELEMENT ZR   BLANK                     0.0000E+00  0.0000E+00  0.0000E+00!
        SPECIES O-2                         O1/-2!
        SPECIES ZR+4                        ZR1/+4!
        PHASE SPINEL %  4 1 2 2 4 !
        CONSTITUENT SPINEL : ZR+4 : VA : VA : O-2 :  !
        PHASE GAS:G %  1  1.0  !
        CONSTITUENT GAS:G :O,ZR :  !
        """
        wks = Workspace(TDB, ['O', 'ZR', 'VA'], ['SPINEL', 'GAS'], {v.P: 1e5, v.X('O'): 1, v.T: 1000})
>       assert np.isnan(wks.get('NP(SPINEL)'))

pycalphad\tests\test_workspace.py:283:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pycalphad\core\workspace.py:461: in get
    result = list(self.get_dict(*args).values())
pycalphad\core\workspace.py:422: in get_dict
    self._expand_property_arguments(args)
pycalphad\core\workspace.py:319: in _expand_property_arguments
    multiplicity = self._detect_phase_multiplicity()
pycalphad\core\workspace.py:305: in _detect_phase_multiplicity
    prop_GM_values = self.eq.GM
pycalphad\core\workspace.py:239: in __get__
    default_value = obj.recompute()
pycalphad\core\workspace.py:299: in recompute
    return _solve_eq_at_conditions(properties, self.phase_record_factory, grid,
pycalphad\core\eqsolver.pyx:242: in pycalphad.core.eqsolver._solve_eq_at_conditions
    result = iter_solver.solve(composition_sets, cur_conds)
pycalphad\core\solver.py:163: in solve
    converged = spec.run_loop(state, 1000)
pycalphad\core\minimizer.pyx:394: in pycalphad.core.minimizer.SystemSpecification.run_loop
    cpdef bint run_loop(self, SystemState state, int max_iterations):
pycalphad\core\minimizer.pyx:403: in pycalphad.core.minimizer.SystemSpecification.run_loop
    eq_soln = solve_state(self, state)
pycalphad\core\minimizer.pyx:833: in pycalphad.core.minimizer.solve_state
    equilibrium_matrix, equilibrium_soln = construct_equilibrium_system(spec, state, 0)
pycalphad\core\minimizer.pyx:684: in pycalphad.core.minimizer.construct_equilibrium_system
    fill_equilibrium_system(equilibrium_matrix, equilibrium_rhs, spec, state)
pycalphad\core\minimizer.pyx:268: in pycalphad.core.minimizer.fill_equilibrium_system
    write_row_fixed_mole_fraction(equilibrium_matrix[component_row_offset + fixed_molefrac_cond_idx, :],
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   (phase_amt[idx]/current_system_amount) * mass_jac[component_idx, num_statevars+j] * c_component[chempot_idx, j]
E   ZeroDivisionError: float division

pycalphad\core\minimizer.pyx:145: ZeroDivisionError
=============================================== short test summary info ===============================================
FAILED pycalphad/tests/test_workspace.py::test_issue_503_pure_vacancy_charge_balance - ZeroDivisionError: float division
================================================== 1 failed in 1.81s ==================================================
```

**Comment by richardotis** [comment]
@nury12n The test will be fixed in #561, though we aren't addressing the underlying numerical issue yet.
